sockeye/training.py

```diff
@@ -341,39 +511,187 @@ def _forward_backward(self, batch: data_io.Batch):
+            )
+        max_out_length = mx.nd.array(max_output_lengths, ctx=translator.context, dtype='int32')
+        if tagged_bt:
+            max_out_length -= 1
+
+        bt_src_lang_nd = mx.nd.full(shape=(1,), val=translator.lang_vocab[bt_src_lang], ctx=translator.context)
+        bt_trg_lang_nd = mx.nd.full(shape=(1,), val=translator.lang_vocab[bt_trg_lang], ctx=translator.context)
+
+        batch_size = batch.data.shape[0]
+        raw_constraints = [None] * batch_size
+        raw_avoid_list = [None] * batch_size
+
+        data = batch.data.as_in_context(translator.context)
+        data_length = batch.data_length.as_in_context(translator.context)
+
+        all_translations = []
+        curr_idx = 0
+        while curr_idx < data.shape[0]:
+            translations = translator._translate_nd(
+                data[curr_idx:min(curr_idx+batch_size, data.shape[0])],
+                data_length[curr_idx:min(curr_idx+batch_size, data.shape[0])],
+                None,
+                raw_constraints,
+                raw_avoid_list,
+                max_out_length,
+                # We translate from target -> source
+                source_lang=bt_trg_lang_nd,
+                target_lang=bt_src_lang_nd,
+            )
+            all_translations.extend(translations)
+            curr_idx += batch_size
+        assert len(all_translations) == data.shape[0]
+
+        max_len = max(len(translation.target_ids) for translation in all_translations)
+
+        source = mx.nd.full(shape=(data.shape[0] * max_len), val=C.PAD_ID, dtype=np.int32)
+
+        words_flat = []
+        positions_flat = []
+        for i, translation in enumerate(all_translations):
+            # words_flat.extend(translation.target_ids)
+            # TODO: properly support target factors
+            words_flat.extend([t[0] for t in translation.target_ids])
+            positions_flat.extend([i * max_len + j for j in range(0, len(translation.target_ids))])
+        words_flat = mx.nd.array(words_flat, dtype=np.int32)
+        positions_flat = mx.nd.array(positions_flat, dtype=np.int32)
+        source[positions_flat] = words_flat
+        source = source.reshape((data.shape[0], max_len, 1))
+
+        if tagged_bt:
+            # Note: we tag with the BOS_SYMBOL token because we know it's in the vocab. This is of course not optimal
+            # and ideally we'd have a special tagging token.
+            tagged = mx.nd.full(shape=(source.shape[0], 1, 1), val=C.BOS_ID, dtype=source.dtype)
+            source = mx.nd.concat(tagged, source, dim=1)
+
+        label = batch.data.squeeze().astype(np.int32)
+        target = label[:, :-1]
+        target = mx.nd.where(target == C.EOS_ID, mx.nd.zeros_like(target), target)  # replace other <eos>'s with <pad>
+        # target: (batch_size, seq_len, 1)
+        target = mx.nd.concat(mx.nd.full((target.shape[0], 1), val=C.BOS_ID, dtype=np.int32), target)
+
+        # TODO: full target factor support
```

Comment:
For my understanding: for bi-directional models that use both source and target factors and in case of backtranslation, the gold target factors should act as source factors for the backtranslation, correct?